### PR TITLE
GH-36730: [Python] Add support for Cython 3.0.0

### LINKED ISSUE #36730: [CI][Python] Cython 3.0 seems to make our build of pyarrow fail
### Describe the bug, including details regarding any error messages, version, and platform.

As seen on some jobs for 13.0.0 on the maintenance branch:
https://github.com/ursacomputing/crossbow/actions/runs/5576946887/jobs/10192099702
and this other branch from main:
https://github.com/ursacomputing/crossbow/actions/runs/5577773946/jobs/10191016009

It seems that the new version of Cython being used (it was released ~1 hour ago at the moment of creation of the issue):
```
 Collecting cython>=0.29.31 (from -r /Users/runner/work/crossbow/crossbow/arrow/python/requirements-wheel-build.txt (line 1))
  Obtaining dependency information for cython>=0.29.31 from https://files.pythonhosted.org/packages/a7/9d/643b508fc18fe095a8c1efea5bd86054ac63974ecd566dba962de40ebaa6/Cython-3.0.0-cp39-cp39-macosx_10_9_x86_64.whl.metadata
  Downloading Cython-3.0.0-cp39-cp39-macosx_10_9_x86_64.whl.metadata (3.1 kB)
```
There is a build issue:
```
 [15/129] cd /Users/runner/work/crossbow/crossbow/arrow/python/build/temp.macosx-10.14-x86_64-cpython-39 && /Users/runner/work/crossbow/crossbow/build-env/bin/python -m cython --cplus --warning-errors --directive embedsignature=True --working /Users/runner/work/crossbow/crossbow/arrow/python --output-file /Users/runner/work/crossbow/crossbow/arrow/python/build/temp.macosx-10.14-x86_64-cpython-39/_flight.cpp /Users/runner/work/crossbow/crossbow/arrow/python/pyarrow/_flight.pyx
FAILED: CMakeFiles/_flight_pyx /Users/runner/work/crossbow/crossbow/arrow/python/build/temp.macosx-10.14-x86_64-cpython-39/CMakeFiles/_flight_pyx 
cd /Users/runner/work/crossbow/crossbow/arrow/python/build/temp.macosx-10.14-x86_64-cpython-39 && /Users/runner/work/crossbow/crossbow/build-env/bin/python -m cython --cplus --warning-errors --directive embedsignature=True --working /Users/runner/work/crossbow/crossbow/arrow/python --output-file /Users/runner/work/crossbow/crossbow/arrow/python/build/temp.macosx-10.14-x86_64-cpython-39/_flight.cpp /Users/runner/work/crossbow/crossbow/arrow/python/pyarrow/_flight.pyx

Error compiling Cython file:
------------------------------------------------------------
...

    cdef cppclass CFlightListing" arrow::flight::FlightListing":
        CResult[unique_ptr[CFlightInfo]] Next()

    cdef cppclass CSimpleFlightListing" arrow::flight::SimpleFlightListing":
        CSimpleFlightListing(vector[CFlightInfo]&& info)
                            ^
------------------------------------------------------------

pyarrow/includes/libarrow_flight.pxd:175:28: Rvalue-reference as function argument not supported

Error compiling Cython file:
------------------------------------------------------------
...
            # Text header values in gRPC (and HTTP/1, HTTP/2) are
            # required to be valid ASCII. Binary header values are
            # exposed as bytes.
            value = value.decode("ascii")
        headers.setdefault(header, []).append(value)
        postincrement(header_iter)
        ^
------------------------------------------------------------

pyarrow/_flight.pyx:2312:8: No 'operator++(int)' declared for postfix '++' (operand type is 'const_iterator')
```

### Component(s)

Continuous Integration, Python

### PR DESCRIPTION
**Pull request by danepitkin**
### Rationale for this change

Cython 3.0.0 is the latest release. PyArrow should work with Cython 3.0.0. **Cython 3 is not enabled in this diff.**

### What changes are included in this PR?

* Don't use `vector[XXX]&&`
* Add a declaration for `postincrement`
  * See also: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#c-postincrement-postdecrement-operator
* Ignore `C4551` warning (function call missing argument list) with MSVC
  * See also: https://github.com/cython/cython/issues/4445
* Add missing `const` to `CLocation`'s static methods.
* Don't use `StopIteration` to stop generator
  * See also: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#python-3-syntax-semantics
* non-extern `cdef` functions will now propagate python exceptions automatically unless explicitly labeled `noexcept`
* Function binding in cython is now enabled by default. Class methods that are used as wrappers for pickling should be converted to staticmethods.
* Numpydocs now validates more Cython 3 objects than Cython <3
  * Enum types are now being validated, and some unhelpful validation checks on Enums are now ignored
* Added a cython <3 nightly CI job

Note:
* Cython 3.0.0, 3.0.1, 3.0.2 has an issue when compiling with debug mode https://github.com/cython/cython/issues/5552

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.

* Closes: #36730

### CODE REVIEW

File: python/pyarrow/tests/parquet/test_dataset.py
Comment (danepitkin) [inline python/pyarrow/tests/parquet/test_dataset.py:1567]: TODO: unskip test before merge

**Comment by danepitkin** [comment]
Potentially related cloudpickle issue? https://github.com/cloudpipe/cloudpickle/issues/506

**Comment by kou** [comment]
I've copied #36745 PR description.
We should update the PR description before we merge this PR.

**Comment by danepitkin** [comment]
@github-actions crossbow submit *python*

**Comment by danepitkin** [comment]
`test-cuda-python` and `test-ubuntu-20.04-python-3` are failing due to this cython issue https://github.com/cython/cython/issues/5552

**Comment by danepitkin** [comment]
Current summary:

* Numpydocs now correctly reports 72 cython functions are missing documentation for their parameters (Should be fixed in a separate PR IMO)
* https://github.com/cython/cython/issues/5552 is breaking `test-cuda-python` and `test-ubuntu-20.04-python-3` nightly jobs
* Dask/Spark failures are on main and can be ignored here

Other than the above issues, I think this PR is ready for review.

**Comment by kou** [comment]
> Numpydocs now correctly reports 72 cython functions are missing documentation for their parameters (Should be fixed in a separate PR IMO)

Could you work on this before we merge this to avoid CI failures on main?

File: python/pyarrow/_dataset_parquet.pyx
Comment (kou) [inline python/pyarrow/_dataset_parquet.pyx:748]: Can we check that this change still works with Cython < 3?

Comment (danepitkin) [reply python/pyarrow/_dataset_parquet.pyx:748]: That reminds me, I wanted to add a Cython<3 nightly job to CI to test future changes don't break anything.

File: python/pyarrow/_dataset_parquet.pyx
Comment (pitrou) [inline python/pyarrow/_dataset_parquet.pyx:747]: Would it work if you make `_reconstruct` above a `staticmethod` instead?

Comment (pitrou) [reply python/pyarrow/_dataset_parquet.pyx:747]: Or it might also be possible to disable the `binding` option specifically for `_reconstruct`:
https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#locally

File: python/pyarrow/_substrait.pyx
Comment (pitrou) [inline python/pyarrow/_substrait.pyx:31]: Are we sure this function cannot raise a Python exception? Its code is definitely non-trivial, so the "solution" here looks more like a bandaid.

In the interest of moving this forward, can you open a bug for this and we'll revisit later?

File: python/pyarrow/scalar.pxi
Comment (pitrou) [inline python/pyarrow/scalar.pxi:813]: Nice catch... I guess this code path is not tested? Can you add a test for it?

File: python/pyarrow/tests/test_fs.py
Comment (pitrou) [inline python/pyarrow/tests/test_fs.py:448]: We have other pickling tests which could benefit from this. Perhaps share this fixture with all test modules and make sure it gets used everywhere possible?
(also perhaps rename it to `picklemodule` for less confusion, though YMMV)

File: python/pyarrow/_dataset_parquet.pyx
Comment (danepitkin) [reply python/pyarrow/_dataset_parquet.pyx:747]: `_reconstruct` as a staticmethod failed with both builtin pickle and cloudpickle, but I can't remember the exact error. I can grab it for you if you'd like. You're right that we should also be able to add a cython decorator to disable binding if that's preferred. I'll go ahead and update with this change.

File: python/pyarrow/_substrait.pyx
Comment (danepitkin) [reply python/pyarrow/_substrait.pyx:31]: This function executes user-defined python code, which absolutely can raise an exception. In fact, the test cases specifically _do_ raise an exception. The problem is that the C++ functionality explicitly chooses to ignore it and return its own error. The pyarrow binding is designed with this in mind, so adding `noexcept` is mimicking the expected behavior of previous cython versions. I'll file an issue to request refactoring of this feature.

File: python/pyarrow/tests/test_fs.py
Comment (danepitkin) [reply python/pyarrow/tests/test_fs.py:448]: Will do! I was thinking the same, but thought it might be better in a different PR. Now that you explicitly mention it, it is probably best to test cloudpickle everywhere as part of this cython upgrade.

File: python/pyarrow/_dataset_parquet.pyx
Comment (pitrou) [reply python/pyarrow/_dataset_parquet.pyx:747]: Which docs? Cython docs?

It works fine in pure Python, but it may be a Cython wart, in which case it would be nice to report an issue:
```python
>>> class C:
...:     @staticmethod
...:     def sm(): pass
...: 
>>> pickle.dumps(C.sm)
b'\x80\x04\x95\x15\x00\x00\x00\x00\x00\x00\x00\x8c\x08__main__\x94\x8c\x04C.sm\x94\x93\x94.'
>>> pickle.loads(pickle.dumps(C.sm))
<function __main__.C.sm()>
>>> pickle.loads(pickle.dumps(C().sm))
<function __main__.C.sm()>
```

**Comment by danepitkin** [comment]
> > Numpydocs now correctly reports 72 cython functions are missing documentation for their parameters (Should be fixed in a separate PR IMO)
> 
> Could you work on this before we merge this to avoid CI failures on main?

Created https://github.com/apache/arrow/pull/37218 ! This fixes most of the issues. I figured we can merge that, rebase here, and then fix anything leftover, which will be manageable to do in this PR. We won't be able to verify the numpydocs test in that PR because it doesn't have cython 3. I tested it with cython 3 locally instead, but didn't see the exact same number of issues that are reported here.

File: python/pyarrow/_dataset_parquet.pyx
Comment (danepitkin) [reply python/pyarrow/_dataset_parquet.pyx:748]: BTW, I did confirm this works on Cython 0.29.36

File: python/pyarrow/_dataset_parquet.pyx
Comment (danepitkin) [reply python/pyarrow/_dataset_parquet.pyx:747]: Nevermind.. it is possible to make this a staticmethod. I don't know why I thought it wasn't working prior. Will update.

File: python/pyarrow/_substrait.pyx
Comment (danepitkin) [reply python/pyarrow/_substrait.pyx:31]: Filed https://github.com/apache/arrow/issues/37235

**Comment by danepitkin** [comment]
@github-actions crossbow submit test-conda-python-3.11-hypothesis

**Comment by danepitkin** [comment]
@github-actions crossbow submit test-conda-python-3.9-pandas-latest

File: python/pyarrow/_dataset_parquet.pyx
Comment (danepitkin) [reply python/pyarrow/_dataset_parquet.pyx]: Ahh I remember what happened. This works for Cython 3, but not Cython 0.29.36.

With Cython 0.29.36, we just need to add the `binding=True` compiler option.

my_test.pyx
```
# cython : language_level = 3
import cython

cdef class Foo():
    @staticmethod
    @cython.binding(True)  # This works
    def _reconstruct():
        return Foo()

    def __reduce__(self):
        return Foo._reconstruct, tuple()


@staticmethod
def _reconstruct(x):
    return Bar(x)


cdef class Bar():
    @staticmethod  # This does not work
    def _reconstruct():
        return Bar()

    def __reduce__(self):
        return Bar._reconstruct, tuple()
```

Usage here:
```
In [1]: from my_test import Foo, Bar

In [2]: f = Foo()

In [3]: b = Bar()

In [4]: import pickle

In [5]: pickle.dumps(f)
Out[5]: b'\x80\x04\x95#\x00\x00\x00\x00\x00\x00\x00\x8c\x07my_test\x94\x8c\x10Foo._reconstruct\x94\x93\x94)R\x94.'

In [6]: pickle.dumps(b)
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
Cell In[6], line 1
----> 1 pickle.dumps(b)

PicklingError: Can't pickle <built-in function _reconstruct>: it's not the same object as my_test._reconstruct
```

File: python/pyarrow/tests/test_fs.py
Comment (danepitkin) [reply python/pyarrow/tests/test_fs.py]: The commit to implement this is pretty big (I already pushed it to this PR). I went ahead and implemented in a separate PR and can rebase on top of it after it merges. This has the additional benefit of testing the CI process with Cython 0.29.X, since I haven't added the Cython < 3 CI job yet.

https://github.com/apache/arrow/pull/37255

**Comment by danepitkin** [comment]
I'd like to merge https://github.com/apache/arrow/pull/37255 and rebase to make this PR cleaner.

**Comment by danepitkin** [comment]
I'm also not sure how to fix these new numpydoc lint issues:

```
INFO:archery:Running Python docstring linters

pyarrow._s3fs.S3LogLevel
-> pyarrow._s3fs.An enumeration.
PR01: Parameters {'qualname', 'module', 'names', 'start', 'type', 'value'} not documented

pyarrow._fs.FileType
-> pyarrow._fs.An enumeration.
PR01: Parameters {'qualname', 'module', 'names', 'start', 'type', 'value'} not documented

pyarrow.lib.MetadataVersion
-> pyarrow.lib.An enumeration.
PR01: Parameters {'qualname', 'module', 'names', 'start', 'type', 'value'} not documented

pyarrow._dataset.type.FileSystemDataset.from_paths
-> pyarrow._dataset.FileSystemDataset.from_paths(cls, paths, schema=None, format=None, filesystem=None, partitions=None, root_partition=None)
PR01: Parameters {'schema', 'filesystem', 'format', 'root_partition', 'partitions'} not documented

pyarrow.lib.MetadataVersion
-> pyarrow.lib.An enumeration.
PR01: Parameters {'qualname', 'module', 'names', 'start', 'type', 'value'} not documented

Total number of docstring violations: 5
```

The enums are all defined in Cython with `cpdef`, which is a wrapper for https://peps.python.org/pep-0435/. Maybe cython or numpydoc isn't parsing correctly?

The `pyarrow._dataset.type.FileSystemDataset.from_paths` error seems like a bug in cython or numpydocs. The parameters are already documented and visible here https://arrow.apache.org/docs/python/generated/pyarrow.dataset.FileSystemDataset.html#pyarrow.dataset.FileSystemDataset.from_paths.

File: python/pyarrow/_s3fs.pyx
Comment (danepitkin) [inline python/pyarrow/_s3fs.pyx:391]: TODO: I'm missing the @staticmethod keyword here

**Comment by danepitkin** [comment]
@github actions crossbow submit test-conda-python-3.11-cython2

**Comment by danepitkin** [comment]
@github-actions crossbow submit test-conda-python-3.10-cython2 --no-cache

**Comment by danepitkin** [comment]
@github-actions archery docker run --no-leaf-cache conda-python-cython2

**Comment by danepitkin** [comment]
@github-actions crossbow submit test-conda-python-3.10-cython2

File: python/pyarrow/_dataset.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:1841]: Was this needed to get some test passing? It seems strange (or wrong?) that this needs to be sorted

Comment (danepitkin) [reply python/pyarrow/_dataset.pyx:1841]: Yes, the `__repr__` test was failing. I thought it might be nice to have deterministic output for the `__repr__`. The key/value pair ordering was swapped when I compiled with Cython 3. Would it be better to update the test case to handle non-determinism?

https://github.com/apache/arrow/blob/main/python/pyarrow/tests/test_dataset.py#L1614
```
def test_fragments_repr(tempdir, dataset):
    # partitioned parquet dataset
    fragment = list(dataset.get_fragments())[0]
    assert (
        repr(fragment) ==
        "<pyarrow.dataset.ParquetFileFragment path=subdir/1/xxx/file0.parquet "
        "partition=[key=xxx, group=1]>"
    )
```

**Review by danepitkin — COMMENTED** [review]
Rebased

File: python/pyarrow/_substrait.pyx
Comment (danepitkin) [reply python/pyarrow/_substrait.pyx:32]: I'll see if I can fix this in a separate PR first.

File: python/pyarrow/_dataset.pyx
Comment (danepitkin) [reply python/pyarrow/_dataset.pyx:1841]: I'll go ahead and remove sorting here and have the test handle non-deterministic repr output

**Comment by danepitkin** [comment]
@github-actions crossbow submit *python*

File: python/pyarrow/_substrait.pyx
Comment (jorisvandenbossche) [reply python/pyarrow/_substrait.pyx:32]: > I'll see if I can fix this in a separate PR first.

It would certainly be nice to fix that, but FWIW I don't think that should hold up this PR even longer

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
> `test-cuda-python` and `test-ubuntu-20.04-python-3` are failing due to this cython issue [cython/cython#5552](https://github.com/cython/cython/issues/5552)

Those are still failing. What do we do about those? Can we pin cython to cython<3 on those builds for now? Or disable gdb support on those builds?

**Comment by danepitkin** [comment]
> > `test-cuda-python` and `test-ubuntu-20.04-python-3` are failing due to this cython issue [cython/cython#5552](https://github.com/cython/cython/issues/5552)
> 
> Those are still failing. What do we do about those? Can we pin cython to cython<3 on those builds for now? Or disable gdb support on those builds?

Good question. Hmm. These are great suggestions. I like the idea of pinning to cython<3 so one can still use gdb. Maybe I can attempt the cython fix later on (and I can file an issue in Arrow to unpin cython for these builds.

File: python/pyarrow/includes/libarrow_fs.pxd
Comment (danepitkin) [reply python/pyarrow/includes/libarrow_fs.pxd]: I'm still not able to reproduce these locally, but I did seem to add them to the wrong place. I updated the code and will test via CI jobs for now.

Comment (danepitkin) [reply python/pyarrow/includes/libarrow_fs.pxd]: Ugh, numpydoc doesn't parse `cdef` and `cpdef` tokens. Only `def` and `class`.

Comment (danepitkin) [reply python/pyarrow/includes/libarrow_fs.pxd]: Hmm, the `cpdef enum` objects are all of type `flag`. 

```
<flag 'S3LogLevel'>
3
6
0
4
1
2
5
<class 'pyarrow._s3fs.S3FileSystem'>
<attribute 'region' of 'pyarrow._s3fs.S3FileSystem' objects>
<class 'pyarrow._fs.PyFileSystem'>
<attribute 'handler' of 'pyarrow._fs.PyFileSystem' objects>
<class 'pyarrow._fs.LocalFileSystem'>
<class 'pyarrow._hdfs.HadoopFileSystem'>
<cyfunction HadoopFileSystem.from_uri at 0x114c41b10>
<class 'pyarrow._gcsfs.GcsFileSystem'>
<attribute 'project_id' of 'pyarrow._gcsfs.GcsFileSystem' objects>
<attribute 'default_bucket_location' of 'pyarrow._gcsfs.GcsFileSystem' objects>
<flag 'FileType'>
<class 'pyarrow._fs.FileSystemHandler'>
<cyfunction FileSystemHandler.open_output_stream at 0x114c7cba0>
```

Comment (danepitkin) [reply python/pyarrow/includes/libarrow_fs.pxd]: I don't think the above is relevant. There just seem to be different types of python objects that get created in cython 2 vs 3 and some of them get parsed differently.

Comment (danepitkin) [reply python/pyarrow/includes/libarrow_fs.pxd]: > BTW, on the latest commit, `FileSystemDataset.from_paths` isn't erroring anymore, so maybe the change from classmethod to staticmethod "fixed" it ..

I noticed that, too! There's some weird stuff going on here with parsing. This change shouldn't have mattered.

For the enums, it seems numpydoc just doesn't parse properly. It's getting the Enum class function signature, but it should really be documenting the values. I don't see any reference to the word "enum" in numpydocs documentation so I'll assume it is unsupported. I can skip enum validation in archery numpydoc. 

```
>>> inspect.signature(pa.MetadataVersion)
<Signature (value, names=None, *, module=None, qualname=None, type=None, start=1, boundary=None)>
>>> inspect.signature(Flag)
<Signature (value, names=None, *, module=None, qualname=None, type=None, start=1, boundary=None)>
```

Comment (danepitkin) [reply python/pyarrow/includes/libarrow_fs.pxd]: Yep, Enums are treated like classes (no special handling) https://github.com/numpy/numpydoc/issues/458

I'll just skip them in archery for now.

**Comment by danepitkin** [comment]
@github-actions crossbow submit *python*

**Comment by danepitkin** [comment]
@github-actions crossbow submit test-cuda-python

**Comment by danepitkin** [comment]
@github-actions crossbow submit *python*

**Comment by danepitkin** [comment]
@github-actions crossbow submit -g python

**Comment by danepitkin** [comment]
Rebased. I am going to wait for Cython 3.0.3, which will fix the `cython --gdb` compilation issue seen in some of the nightly builds.

**Comment by danepitkin** [comment]
@github-actions crossbow submit -g python

**Comment by danepitkin** [comment]
I've separated out the enablement of Cython 3 into a separate diff: https://github.com/apache/arrow/pull/37743

Now, we can merge Cython 3 support and then enable Cython 3 once Cython 3.0.3 is released.

File: python/pyarrow/_substrait.pyx
Comment (danepitkin) [reply python/pyarrow/_substrait.pyx:32]: FYI I haven't attempted #37235. I hope its okay to merge as-is given how long this PR has taken.

**Comment by danepitkin** [comment]
IMO this can be merged now!

File: python/pyarrow/_dataset.pyx
Comment (pitrou) [inline python/pyarrow/_dataset.pyx:1080]: Hmm... classmethod doesn't work correctly on Cython 3 anymore? Is it a known issue? Or is this change actually not needed?

File: python/pyarrow/_substrait.pyx
Comment (pitrou) [reply python/pyarrow/_substrait.pyx:32]: Can you just add a comment pointing to GH-37235?

File: python/pyarrow/_dataset.pyx
Comment (danepitkin) [reply python/pyarrow/_dataset.pyx:1080]: classmethod does work, I can change it back. I tried switching to staticmethod when numpydocs were failing, but it turns out it was the comment that was causing numpydoc parsing errors. I thought staticmethod was a slight improvement since `cls` wasn't actually used in the classmethod.

File: python/pyarrow/_substrait.pyx
Comment (danepitkin) [reply python/pyarrow/_substrait.pyx:32]: Will do!

**Review by AlenkaF — APPROVED** [review]
Thank you for working on this Dane!! Great work 👍

**Comment by AlenkaF** [comment]
@github-actions crossbow submit -g python

**Comment by AlenkaF** [comment]
@github-actions crossbow submit *python*

**Comment by kou** [comment]
We need to fix #37803 to use `@github-actions crossbow submit ...`. :<

**Comment by AlenkaF** [comment]
> We need to fix #37803 to use `@github-actions crossbow submit ...`. :<

Theaks! Got it now =)

**Comment by AlenkaF** [comment]
@github-actions crossbow submit -g python

**Comment by AlenkaF** [comment]
None of the failures look related, wil merge. Thanks again Dane!

**Pull request merged**
